File: backend/src/main/java/by/fpmibsu/findafriend/controller/controllers/UserShelterController.java
package by.fpmibsu.findafriend.controller.controllers;

import by.fpmibsu.findafriend.application.HandleResult;
import by.fpmibsu.findafriend.application.controller.*;
import by.fpmibsu.findafriend.application.mediatr.Mediatr;
import by.fpmibsu.findafriend.controller.AuthUtils;
import by.fpmibsu.findafriend.controller.Logging;
import by.fpmibsu.findafriend.controller.commands.usershelter.AddUserToShelterCommand;
import by.fpmibsu.findafriend.controller.commands.usershelter.RemoveUserFromShelterCommand;
import by.fpmibsu.findafriend.entity.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@ControllerRoute(route = "/userShelter")
public class UserShelterController extends Controller {
    private static final Logger logger = LogManager.getLogger(UserController.class);
    private final Mediatr mediatr;

    public UserShelterController(Mediatr mediatr) {
        this.mediatr = mediatr;
    }

    @RequireAuthentication
    @Endpoint(path = "/addUserToShelter", method = HttpMethod.POST)
    public HandleResult addUserToShelter(@FromQuery(parameterName = "shelterId") int shelterId,
                                         @FromQuery(parameterName = "userId") int userId,
                                         @WebToken(parameterName = "role") String role) throws Exception {
        if (!AuthUtils.allowRoles(role, User.Role.MODERATOR, User.Role.ADMINISTRATOR)) {
            Logging.warnNonAuthorizedAccess(this.request, logger);
            return notAuthorized();
        }

        return ok(mediatr.send(new AddUserToShelterCommand(userId, shelterId)));
    }

    @RequireAuthentication
    @Endpoint(path = "/removeUserFromShelter", method = HttpMethod.POST)
    public HandleResult removeFromShelter(@FromQuery(parameterName = "userId") int userId,
                                          @WebToken(parameterName = "role") String role) throws Exception {
        if (!AuthUtils.allowRoles(role, User.Role.MODERATOR, User.Role.ADMINISTRATOR)) {
            Logging.warnNonAuthorizedAccess(this.request, logger);
            return notAuthorized();
        }

        return ok(mediatr.send(new RemoveUserFromShelterCommand(userId)));
    }
}

File: BackendNET/Backend/Controllers/UserShelterController.cs
using Backend.Commands.UserShelter.AddUserToShelter;
using Backend.Commands.UserShelter.RemoveUserFromShelter;
using MediatR;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Backend.Controllers;

[Authorize(Roles = "Administrator,Moderator")]
public class UserShelterController : ControllerBase
{
    private readonly IMediator _mediator;

    public UserShelterController(IMediator mediator)
    {
        _mediator = mediator;
    }

    [HttpPost("/userShelter/addUserToShelter")]
    public async Task<IActionResult> AddUserToShelterAsync([FromQuery] int shelterId, [FromQuery] int userId)
    {
        return Ok(await _mediator.Send(new AddUserToShelterCommand { UserId = userId, ShelterId = shelterId }));
    }

    [HttpPost("/userShelter/removeUserFromShelter")]
    public async Task<IActionResult> RemoveFromShelterAsync([FromQuery] int userId)
    {
        return Ok(await _mediator.Send(new RemoveUserFromShelterCommand { UserId = userId }));
    }
}